### PR TITLE
TYP: remove mypy ignore from pandas/core/construction.py

### PR DESCRIPTION
**Pull request by natmokval**
Related to #37715
mypy ignore[assignment] was removed from pandas/core/construction.py

### CODE REVIEW

File: pandas/core/construction.py [hunk 1]
```diff
@@ -502,9 +502,7 @@ def sanitize_masked_array(data: ma.MaskedArray) -> np.ndarray:
     if mask.any():
         dtype, fill_value = maybe_promote(data.dtype, np.nan)
         dtype = cast(np.dtype, dtype)
-        # Incompatible types in assignment (expression has type "ndarray[Any,
-        # dtype[Any]]", variable has type "MaskedArray[Any, Any]")
-        data = data.astype(dtype, copy=True)  # type: ignore[assignment]
+        data = ma.asarray(data.astype(dtype, copy=True))
```
Comment (MarcoGorelli) [inline pandas/core/construction.py:505]: thanks for your PR - would it work to just rename `data`? looks like the issue that it's getting reassigned to a different type

something like
```
    data_newname = data.astype(dtype, copy=True)
    data_newname.soften_mask()
    data_newname[mask] = fill_value
    return data_newname
else:
    return data.copy()
```
(the hardest part would arguably be coming up with a good name, i.e. not `data_newname` but something more descriptive, but I haven't looked at this carefully enough yet to think of what)

Comment (natmokval) [reply pandas/core/construction.py:505]: Thank you, @MarcoGorelli for your comment. I will do, as you suggest. What do you think, could be `sanitized_data` or `sanitized_ma` a suitable new name for `data`?

**Review by MarcoGorelli — APPROVED** [review]
looks good to me, leaving open a bit in case others have comments

EDIT: oops, not green

**Comment by natmokval** [comment]
Looks like mypy raises an error, I will fix it.

**Comment by MarcoGorelli** [comment]
I think the issue might be that `astype` isn't meant to be called on `numpy.ma.MaskedArray` https://numpy.org/doc/stable/reference/routines.ma.html

Calling it just returns `ndarray`:
```python
data = np.ma.MaskedArray([1,2,3], mask=[False, False, True])
reveal_type(data)  # Revealed type is "numpy.ma.core.MaskedArray[Any, Any]"
reveal_type(data.astype('int64', copy=True))  # Revealed type is "numpy.ndarray[Any, numpy.dtype[Any]]"
```
Might be worth checking if there's an issue in numpy about it, and if not, then reporting to them?

The method which says "Convert the input to a masked array of the given data-type" is the one you'd originally used: `numpy.ma.asarray`, so your original solution was probably the correct one, sorry for the noise 🙈 The docs say it that `asarray` doesn't copy if the original data is an ndarray anyway

**Comment by natmokval** [comment]
> I think the issue might be that `astype` isn't meant to be called on `numpy.ma.MaskedArray` https://numpy.org/doc/stable/reference/routines.ma.html
> 
> Calling it, the return type is typed to be `ndarray`, even though if we print it we see it's `numpy.ma.MaskedArray`:
> 
> ```python
> data = np.ma.MaskedArray([1,2,3], mask=[False, False, True])
> reveal_type(data)  # Revealed type is "numpy.ma.core.MaskedArray[Any, Any]"
> reveal_type(data.astype('int64', copy=True))  # Revealed type is "numpy.ndarray[Any, numpy.dtype[Any]]"
> ```
> 
> Might be worth checking if there's an issue in numpy about it, and if not, then reporting to them? Perhaps this could be made more precise on their end?

Thanks, I'll check if there's an issue on GitHub about it.
> 
> The method which says "Convert the input to a masked array of the given data-type" is the one you'd originally used: `numpy.ma.asarray`, so your original solution was probably the correct one, sorry for the noise 🙈 The docs say it that `asarray` doesn't copy if the original data is an ndarray anyway

I rolled back the commit to the first one.

**Review by MarcoGorelli — APPROVED** [review]
let's get this in, thanks @natmokval

**Pull request merged**